File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque

def bfs(maps, start, end):
    
    rows = len(maps)
    cols = len(maps[0])
    
    queue = deque([(start, 0)]) # (start, 거리)
    visited = [[False] * cols for _ in range(rows)]
    visited[start[0]][start[1]] = True

    while queue:
        (x, y), distance = queue.popleft()
        
        if (x, y) == end:
            return distance
        
        for dx, dy in[(-1, 0), (1, 0), (0, -1), (0, 1)]:
            nx = x + dx
            ny = y + dy
            if 0 <= nx < rows and 0 <= ny < cols and maps[nx][ny] != 'X' and not visited[nx][ny]:
                visited[nx][ny] = True
                queue.append(((nx, ny), distance + 1))
    return -1

def solution(maps):
    
    for i in range(len(maps)):
        for j in range(len(maps[i])):
            if maps[i][j] == 'S':
                start = (i, j)
            elif maps[i][j] == 'L':
                lever = (i, j)
            elif maps[i][j] == 'E':
                exit = (i, j)
    
    to_lever = bfs(maps, start, lever)
    if to_lever == -1:
        return -1
    
    to_exit = bfs(maps, lever, exit)
    if to_exit == -1:
        return -1
    
    return to_lever + to_exit

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    public int solution(String[] maps) {
        int n = maps.length;
        int m = maps[0].length();
        int startX = 0, startY = 0, leverX = 0, leverY = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                char c = maps[i].charAt(j);
                if (c == 'S') {
                    startX = i;
                    startY = j;
                }
                if (c == 'L') {
                    leverX = i;
                    leverY = j;
                }
            }
        }
        
        int dist1 = bfs(maps, startX, startY, 'L');
        int dist2 = bfs(maps, leverX, leverY, 'E');
        
        if (dist1 == -1 || dist2 == -1)
            return -1;
        else
            return dist1 + dist2;
    }
    
    private int bfs(String[] maps, int startX, int startY, char target) {
        int n = maps.length;
        int m = maps[0].length();
        int[][] distance = new int[n][m];
        boolean[][] visited = new boolean[n][m];
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};

        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{startX, startY});
        visited[startX][startY] = true;
        distance[startX][startY] = 0;

        while (!q.isEmpty()) {
            int[] curr = q.poll();
            int x = curr[0];
            int y = curr[1];

            if (maps[x].charAt(y) == target) {
                return distance[x][y];
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
                    if (!visited[nx][ny] && maps[nx].charAt(ny) != 'X') {
                        visited[nx][ny] = true;
                        distance[nx][ny] = distance[x][y] + 1;
                        q.add(new int[]{nx, ny});
                    }
                }
            }
        }

        return -1;
    }
}